File: matlab/demos/demo_audioAnalysis.m
%% demo_audioAnalysis.m
%  Extract spectral peaks from audio files and compute perceptual features.
%
%  Demonstrates the workflow for analysing real audio: peak extraction
%  via audioPeaks, then spectral similarity (SPCS), harmonicity,
%  spectral entropy, roughness, and virtual pitches — all without
%  spectral enrichment, since the extracted peaks already represent
%  the full spectrum.
%
%  Two peak-extraction passes are shown:
%    1. No smoothing — appropriate for steady-state sounds (piano, oboe)
%       but produces many spurious peaks for sounds with vibrato or
%       frequency jitter (violin, complex music).
%    2. With Gaussian smoothing (sigmaPeaks cents) — collapses
%       vibrato-spread energy into single peaks, giving a cleaner
%       representation for all sources.
%
%  Perceptual features are computed from the smoothed peaks.
%
%  The audio files are in the audio/ subfolder.
%
%  Uses: audioPeaks, convertPitch, spectralEntropy, templateHarmonicity,
%        roughness, virtualPitches, cosSimExpTens
%  (from the Music Perception Toolbox).

%% === User-adjustable parameters ===

sigma      = 12;    % Gaussian smoothing width for perceptual measures
sigmaPeaks = 12;    % Gaussian smoothing width for peak extraction

% Audio files to analyse (relative to the toolbox root)
audioFiles = {
    '../audio/piano_C4.wav',                'Piano C4'
    '../audio/violin_A4.wav',               'Violin A4'
    '../audio/oboe_A4.wav',                 'Oboe A4'
    '../audio/Piano_Emin.wav',              'Piano E minor'
    '../audio/Piano_G7_3rd_inversion.wav',  'Piano G7 (3rd inv.)'
    '../audio/Piano_Cmin_open.wav',         'Piano C minor (open)'
    '../audio/music_sample.wav',            'Music sample'
};

nFiles = size(audioFiles, 1);

%% === Pass 1: Extract peaks without smoothing ===

fprintf('=== Pass 1: Peak extraction (no smoothing) ===\n\n');

rawPeakCounts = zeros(nFiles, 1);
validFiles    = false(nFiles, 1);

for i = 1:nFiles
    filepath = audioFiles{i, 1};
    label    = audioFiles{i, 2};

    if ~isfile(filepath)
        fprintf('  %s: file not found (%s), skipping.\n\n', label, filepath);
        continue;
    end

    [f, ~] = audioPeaks(filepath);
    rawPeakCounts(i) = numel(f);
    validFiles(i)    = true;
    fprintf('\n');
end

%% === Pass 2: Extract peaks with smoothing ===

fprintf('=== Pass 2: Peak extraction (sigma = %d cents) ===\n\n', sigmaPeaks);

results = struct('label', {}, 'f', {}, 'w', {}, 'p', {});

for i = 1:nFiles
    filepath = audioFiles{i, 1};
    label    = audioFiles{i, 2};

    if ~validFiles(i)
        continue;
    end

    [f, w] = audioPeaks(filepath, 'sigma', sigmaPeaks);
    p = convertPitch(f, 'hz', 'cents');

    idx = numel(results) + 1;
    results(idx).label = label;
    results(idx).f = f;
    results(idx).w = w;
    results(idx).p = p;
    results(idx).rawPeakCount = rawPeakCounts(i);

    fprintf('\n');
end

nResults = numel(results);

if nResults == 0
    fprintf('No audio files found. Check the audio/ folder.\n');
    return;
end

%% === Peak count comparison ===

fprintf('=== Peak count comparison ===\n\n');
fprintf('%-25s  %10s  %10s\n', 'Sound', 'Unsmoothed', 'Smoothed');
fprintf('%s\n', repmat('-', 1, 50));

for i = 1:nResults
    fprintf('%-25s  %10d  %10d\n', ...
        results(i).label, results(i).rawPeakCount, numel(results(i).f));
end

%% === Single-file features (smoothed peaks, no spectral enrichment) ===

fprintf('\n=== Single-file features (smoothed peaks, sigma = %d) ===\n\n', sigma);
fprintf('%-25s  %5s  %8s  %8s  %8s  %8s\n', ...
    'Sound', 'Peaks', 'specEnt', 'hMax', 'hEnt', 'Rough');
fprintf('%s\n', repmat('-', 1, 72));

for i = 1:nResults
    f = results(i).f;
    w = results(i).w;
    p = results(i).p;

    H = spectralEntropy(p, w, sigma);
    [hMax, hEnt] = templateHarmonicity(p, w, sigma);
    r = roughness(f, w);

    results(i).specEnt = H;
    results(i).hMax    = hMax;
    results(i).hEnt    = hEnt;
    results(i).rough   = r;

    fprintf('%-25s  %5d  %8.4f  %8.4f  %8.4f  %8.4f\n', ...
        results(i).label, numel(f), H, hMax, hEnt, r);
end

%% === Pairwise spectral pitch class similarity ===

fprintf('\n=== Pairwise SPCS (smoothed peaks, sigma = %d) ===\n\n', sigma);

% Header
fprintf('%25s  ', '');
for j = 1:nResults
    lbl = results(j).label;
    if numel(lbl) > 12, lbl = lbl(1:12); end
    fprintf('%-12s', lbl);
end
fprintf('\n');

for i = 1:nResults
    fprintf('%-25s  ', results(i).label);
    for j = 1:nResults
        if j < i
            fprintf('%12s', '');
        elseif j == i
            fprintf('%-12s', '1.000');
        else
            s = cosSimExpTens(results(i).p, results(i).w, ...
                              results(j).p, results(j).w, ...
                              sigma, 1, false, true, 1200, ...
                              'verbose', false);
            fprintf('%-12.3f', s);
        end
    end
    fprintf('\n');
end

%% === Virtual pitches ===

fprintf('\n=== Strongest virtual pitch per sound (smoothed peaks) ===\n\n');
fprintf('%-25s  %10s  %10s  %10s\n', ...
    'Sound', 'VP (cents)', 'VP (MIDI)', 'Salience');
fprintf('%s\n', repmat('-', 1, 60));

for i = 1:nResults
    [vp_p, vp_w] = virtualPitches(results(i).p, results(i).w, sigma);
    [maxW, maxIdx] = max(vp_w);
    bestCents = vp_p(maxIdx);
    bestMidi  = convertPitch(bestCents, 'cents', 'midi');

    fprintf('%-25s  %10.1f  %10.2f  %10.3f\n', ...
        results(i).label, bestCents, bestMidi, maxW);
end

fprintf('\nDone.\n');

File: python/demos/demo_audio_analysis.py
"""demo_audio_analysis.py

Extract spectral peaks from audio files and compute perceptual features.

Demonstrates the workflow for analysing real audio: peak extraction
via audio_peaks, then spectral similarity (SPCS), harmonicity,
spectral entropy, roughness, and virtual pitches — all without
spectral enrichment, since the extracted peaks already represent
the full spectrum.

Two peak-extraction passes are shown:
  1. No smoothing — appropriate for steady-state sounds (piano, oboe)
     but produces many spurious peaks for sounds with vibrato or
     frequency jitter (violin, complex music).
  2. With Gaussian smoothing (sigma_peaks cents) — collapses
     vibrato-spread energy into single peaks, giving a cleaner
     representation for all sources.

Perceptual features are computed from the smoothed peaks.

The audio files are in the audio/ subfolder (one level up from
demos/).

Requires: soundfile (pip install soundfile)
"""

import numpy as np
from pathlib import Path

import sys, os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))
import mpt

# ===================================================================
#  User-adjustable parameters
# ===================================================================

sigma = 12           # Gaussian smoothing width for perceptual measures
sigma_peaks = 12     # Gaussian smoothing width for peak extraction

# Audio files to analyse (relative to this script's parent directory)
audio_dir = Path(__file__).resolve().parent.parent / 'audio'

audio_files = [
    ('piano_C4.wav',                  'Piano C4'),
    ('violin_A4.wav',                 'Violin A4'),
    ('oboe_A4.wav',                   'Oboe A4'),
    ('Piano_Emin.wav',                'Piano E minor'),
    ('Piano_G7_3rd_inversion.wav',    'Piano G7 (3rd inv.)'),
    ('Piano_Cmin_open.wav',           'Piano C minor (open)'),
    ('music_sample.wav',              'Music sample'),
]

# ===================================================================
#  Pass 1: Extract peaks without smoothing
# ===================================================================

print("=== Pass 1: Peak extraction (no smoothing) ===\n")

raw_results = []
for filename, label in audio_files:
    filepath = audio_dir / filename
    if not filepath.is_file():
        print(f"  {label}: file not found ({filepath}), skipping.")
        continue

    f, w, _ = mpt.audio_peaks(str(filepath))

    raw_results.append({
        'label': label,
        'filename': filename,
        'n_peaks': len(f),
    })
    print()

# ===================================================================
#  Pass 2: Extract peaks with smoothing
# ===================================================================

print(f"=== Pass 2: Peak extraction (sigma = {sigma_peaks} cents) ===\n")

smooth_results = []
for filename, label in audio_files:
    filepath = audio_dir / filename
    if not filepath.is_file():
        continue

    f, w, _ = mpt.audio_peaks(str(filepath), sigma=sigma_peaks)
    p = mpt.convert_pitch(f, 'hz', 'cents')

    smooth_results.append({
        'label': label,
        'f': f,
        'w': w,
        'p': p,
    })
    print()

if len(smooth_results) == 0:
    print("No audio files found. Check the audio/ folder.")
    exit()

# ===================================================================
#  Peak count comparison
# ===================================================================

print("=== Peak count comparison ===\n")
print(f"{'Sound':<25s}  {'Unsmoothed':>10s}  {'Smoothed':>10s}")
print('-' * 50)

for raw, smooth in zip(raw_results, smooth_results):
    print(f"{raw['label']:<25s}  {raw['n_peaks']:10d}  {len(smooth['f']):10d}")

# ===================================================================
#  Single-file features (smoothed peaks, no spectral enrichment)
# ===================================================================

print(f"\n=== Single-file features (smoothed peaks, sigma = {sigma}) ===\n")
print(f"{'Sound':<25s}  {'Peaks':>5s}  {'specEnt':>8s}  {'hMax':>8s}  "
      f"{'hEnt':>8s}  {'Rough':>8s}")
print('-' * 72)

for res in smooth_results:
    f, w, p = res['f'], res['w'], res['p']

    H = mpt.spectral_entropy(p, w, sigma)
    hMax, hEnt = mpt.template_harmonicity(p, w, sigma)
    r = mpt.roughness(f, w)

    res['spec_ent'] = H
    res['h_max'] = hMax
    res['h_ent'] = hEnt
    res['rough'] = r

    print(f"{res['label']:<25s}  {len(f):5d}  {H:8.4f}  {hMax:8.4f}  "
          f"{hEnt:8.4f}  {r:8.4f}")

# ===================================================================
#  Pairwise spectral pitch class similarity
# ===================================================================

print(f"\n=== Pairwise SPCS (smoothed peaks, sigma = {sigma}) ===\n")

labels = [r['label'] for r in smooth_results]
col_w = 12
print(' ' * 26 + ''.join(f"{l[:col_w]:<{col_w}s}" for l in labels))

for i, res_i in enumerate(smooth_results):
    row_str = f"{res_i['label']:<25s} "
    for j, res_j in enumerate(smooth_results):
        if j < i:
            row_str += ' ' * col_w
        elif j == i:
            row_str += f"{'1.000':<{col_w}s}"
        else:
            s = mpt.cos_sim_exp_tens_raw(
                res_i['p'], res_i['w'],
                res_j['p'], res_j['w'],
                sigma, 1, False, True, 1200,
                verbose=False,
            )
            row_str += f"{s:<{col_w}.3f}"
    print(row_str)

# ===================================================================
#  Virtual pitches
# ===================================================================

print(f"\n=== Strongest virtual pitch per sound (smoothed peaks) ===\n")
print(f"{'Sound':<25s}  {'VP (cents)':>10s}  {'VP (MIDI)':>10s}  {'Salience':>10s}")
print('-' * 60)

for res in smooth_results:
    vp_p, vp_w = mpt.virtual_pitches(res['p'], res['w'], sigma)
    max_idx = np.argmax(vp_w)
    best_cents = vp_p[max_idx]
    best_midi = mpt.convert_pitch(best_cents, 'cents', 'midi')

    print(f"{res['label']:<25s}  {best_cents:10.1f}  {best_midi:10.2f}  "
          f"{vp_w[max_idx]:10.3f}")

print("\nDone.")
